src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
+  private void recover(TopicPartition topicPart) throws IOException {
+    if (!recovered.contains(topicPart)) {
+      WAL wal = new WAL(conf, url, topicsDir, topicPart);
+      wal.apply();
+      wals.put(topicPart, wal);
+      recovered.add(topicPart);
+      readOffsets(topicPart);
+    }
+  }
+
+  private void readOffsets() throws IOException {
     Path path = new Path(url + "/" + topicsDir);
-    PathFilter filter = new CommittedFileFilter();
     FileStatus[] topics = fs.listStatus(path);
-    for (FileStatus topic: topics) {
+    for (FileStatus topic : topics) {
       if (topic.isDirectory()) {
         String topicName = topic.getPath().getName();
-        FileStatus[] partitions = fs.listStatus(topic.getPath(), filter);
+        FileStatus[] partitions = fs.listStatus(topic.getPath());
         for (FileStatus partition : partitions) {
-          if (partition.isFile()) {
-            String filename = partition.getPath().getName();
-            String[] parts = filename.split("\\.");
-            try {
-              int partitionNum = Integer.parseInt(parts[0]);
-              long offset = Long.parseLong(parts[1]);
-              TopicPartition topicPart = new TopicPartition(topicName, partitionNum);
-              if (!offsets.containsKey(topicPart) || offset > offsets.get(topicPart)) {
-                offsets.put(topicPart, offset);
-              }
-            } catch (NumberFormatException e) {
-              log.warn("Invalid Copycat HDFS file: {}", filename);
-            }
+          if (partition.isDirectory()) {
+            int partitionNum = Integer.parseInt(partition.getPath().getName());
+            TopicPartition topicPart = new TopicPartition(topicName, partitionNum);
+            readOffsets(topicPart);
           }
         }
       }
     }
-    return offsets;
   }
 
+  private void readOffsets(TopicPartition topicPart) throws IOException {
+    String topic = topicPart.topic();
+    int partition = topicPart.partition();
+    Path path = new Path(url + "/" + topicsDir + "/" + topic + "/" + partition);
+    if (!fs.exists(path)) {
+      return;
+    }
+    PathFilter filter = new CommittedFileFilter();
+    FileStatus[] committedFiles = fs.listStatus(path, filter);
+    for (FileStatus committedFile : committedFiles) {
+      String filename = committedFile.getPath().getName();
+      String[] parts = filename.split("_");
+      try {
+        long endOffset = Long.parseLong(parts[1]);
+        if (!offsets.containsKey(topicPart) || endOffset > offsets.get(topicPart)) {
+          offsets.put(topicPart, endOffset + 1);
+        }
+      } catch (NumberFormatException e) {
+        log.warn("Invalid committed file: {}", filename);
+      }
+    }
+  }
+
+
   private class CommittedFileFilter implements PathFilter {
+
+    @Override
     public boolean accept(Path path) {
-      String[] parts = path.getName().split("\\.");
+      String filename = path.getName();
```

Comment:
It'd be good to have some tests for code like this. Manually parsing strings like this usually has some edge cases that are easy to miss.